core/src/test/java/org/bitcoinj/core/TransactionTest.java

```diff
@@ -305,6 +305,73 @@ public void witnessTransaction() {
                 tx.getHash().toString());
     }
 
+    @Test
+    public void testWitnessSignature() {
+        // test vector P2SH-P2WPKH from:
+        // https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+        NetworkParameters netParams = TestNet3Params.get();
+
+        Transaction unsignedTx = new Transaction(netParams, HEX.decode("01000000" + // version
+                "01" + // input * 1
+                "db6b1b20aa0fd7b23880be2ecbd4a98130974cf4748fb66092ac4d3ceb1a5477" + "01000000" + "00" + "feffffff" +
+                "02" + // output * 2
+                "b8b4eb0b00000000" + "1976a914a457b684d7f0d539a46a45bbc043f35b59d0d96388ac" +
+                "0008af2f00000000" + "1976a914fd270b1ee6abcaea97fea7ad0402e8bd8ad6d77c88ac" +
+                "92040000"));
+        assertEquals(1, unsignedTx.getInputs().size());
+        assertEquals(2, unsignedTx.getOutputs().size());
+
+        ECKey key = ECKey.fromPrivate(
+                HEX.decode("eb696a065ef48a2192da5b28b694f87544b30fae8327c4510137a922f32c6dcf"));
+        assertEquals("03ad1d8e89212f0b92c74d23bb710c00662ad1470198ac48c43f7d6f93a2a26873",
+                key.getPublicKeyAsHex());
+
+        Script redeemScript = ScriptBuilder.createP2WPKHOutputScript(key);
+        assertEquals("001479091972186c449eb1ded22b78e40d009bdf0089",
+                HEX.encode(redeemScript.getProgram()));
+
+        byte[] p2wpkhHash = Utils.sha256hash160(redeemScript.getProgram());
+        Script scriptPubKey = ScriptBuilder.createOutputScript(
+                LegacyAddress.fromScriptHash(netParams, p2wpkhHash));
+        assertEquals("a9144733f37cf4db86fbc2efed2500b4f4e49f31202387",
+                HEX.encode(scriptPubKey.getProgram()));
+
+        Script scriptCode = new ScriptBuilder()
+                .data(ScriptBuilder.createOutputScript(LegacyAddress.fromKey(netParams, key))
+                        .getProgram())
+                .build();
+        assertEquals("1976a91479091972186c449eb1ded22b78e40d009bdf008988ac",
+                HEX.encode(scriptCode.getProgram()));
+
+        TransactionSignature txSig = unsignedTx.calculateWitnessSignature(0, key,
+                scriptCode, Coin.COIN.multiply(10),
+                Transaction.SigHash.ALL, false);
```

Comment:
Thanks for the test. It would be nice if there were tests for other SigHash and anyoneCanPay values. Are there any in Bitcoin Core that could be adapted?

Reply:
Just got through wiki, only basic sample was provided. Bitcoin core should have test cases for other flags in the code.
Maybe in a separate test later?

Reply:
Ok, fair enough. But we should really add more tests before release.